Add support for Clang-Format

agarny opened the pull request:
Addresses issue #284.

Clang-Format is not perfect in that it "only" allows the styling of certain things. So, although this PR tries to stick to our current style, there are some cases where this was just not possible, e.g.
- `Component& operator=(Component m); /**< Assignment operator */`
   has become
   `Component &operator=(Component m); /**< Assignment operator */`
  i.e. there is a space **before** `&` not **after**. The idea was to keep things such as `const std::string &text`, but this means that for methods that return a reference things look a bit strange.
- `bool removeComponent(const std::string &name, bool searchEncapsulated=true);`
  has become
  `bool removeComponent(const std::string &name, bool searchEncapsulated = true);`
  i.e. there is a space **before** and **after** `=`, which I think is completely fine (and what I would have done in the first place), it's just that it's not the style currently used throughout the code. However, there is nothing we can do about this. Clang-Format always puts a space **before** and **after** operators such as `=`.
- `Model(Model &&rhs); /**< Move constructor */`
  has become 
  `Model(Model && rhs); /**< Move constructor */`
  i.e. there is a space **before** and **after** `&&`... since it's an operator, but this obviously means that it now looks a bit strange.
- `const char * const CELLML_2_0_NS = "http://www.cellml.org/cellml/2.0#";`
  has become
  `const char *const CELLML_2_0_NS = "http://www.cellml.org/cellml/2.0#";`
  i.e. there is only a space **before** `*` and none **after**. Again, nothing we can do about this.
- 
```
err->setDescription("Model element is in invalid namespace '" + nodeNamespace +
                    "'. A valid CellML root node should be in namespace '" + CELLML_2_0_NS +
                    "'.");
```
  has become
  `err->setDescription("Model element is in invalid namespace '" + nodeNamespace + "'. A valid CellML root node should be in namespace '" + CELLML_2_0_NS + "'.");`
  i.e. everything on one line.
- `return printComponent(std::shared_ptr<Component>(std::shared_ptr<Component>{}, &component));`
  has become
  `return printComponent(std::shared_ptr<Component>(std::shared_ptr<Component> {}, &component));`
  i.e. there is a space **before** `{}`. Again, nothing we can do about it (or so it seems).
- Things like
```
"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
"<model xmlns=\"http://www.cellml.org/cellml/2.0#\">"
    "<component name=\"parent_component\"/>"
    "<component name=\"child1\"/>"
    "<component name=\"child2\"/>"
    "<component name=\"child3\"/>"
    "<encapsulation>"
        "<component_ref component=\"parent_component\">"
            "<component_ref component=\"child1\"/>"
            "<component_ref component=\"child2\"/>"
            "<component_ref component=\"child3\"/>"
        "</component_ref>"
    "</encapsulation>"
"</model>";
```
have become
```
"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
"<model xmlns=\"http://www.cellml.org/cellml/2.0#\">"
"<component name=\"parent_component\"/>"
"<component name=\"child1\"/>"
"<component name=\"child2\"/>"
"<component name=\"child3\"/>"
"<encapsulation>"
"<component_ref component=\"parent_component\">"
"<component_ref component=\"child1\"/>"
"<component_ref component=\"child2\"/>"
"<component_ref component=\"child3\"/>"
"</component_ref>"
"</encapsulation>"
"</model>";
```
i.e. we have lost our "customised" indentation, which is not surprising. It does, however, raises another issue which is that of our Printer and the fact that we don't indent and use line feeds when printing out a CellML model. This is a different issue though (see issue #309).

agarny committed:
Add support for Clang-Format.

agarny committed:
Add support for Clang-Format.

agarny committed:
Add support for Clang-Format.

agarny committed:
Files formatting changed as a result of Clang-Format (and a bit of manual tweaking).

nickerso commented on CMakeLists.txt:
```diff
@@ -163,3 +163,15 @@ endif()
+
+# Format our code
+if(CLANG_FORMAT_EXE)
+    file(GLOB_RECURSE FILES_TO_FORMAT "${CMAKE_CURRENT_SOURCE_DIR}/src/*.cpp"
```
don't we have variables listing the filenames for all the libCellML source, header files, tests, etc. ? And if so, shouldn't we use those rather than globbing all files, which might include non-libCellML files?

agarny replied:
True, but they are "buried" in sub CMakeLists.txt files. So, I guess we have two choices: 1) keep things as they or 2) call clang-format in all sub CMakeLists.txt files, which is a bit of a pain and something that we might forget to do if we were ever going to add a sub CMakeLists.txt file.

Happy to go with whatever the majority decides though.

hsorby replied:
We could use `PARENT_SCOPE` to pass the full list of the files to format up to this CMakeLists.txt file.

agarny replied:
I am in favour of the `PARENT_SCOPE` approach. Ok, will look into it sometime this week.

nickerso reviewed (commented):
Is clang-format running on all the build slaves? Does it give the same results across all platforms?

agarny commented:
> Is clang-format running on all the build slaves? Does it give the same results across all platforms?

@hsorby to confirm, but to be honest, I am not sure it matters since it's something that should be run on people's local machine, so that they can make the right changes before creating a PR. This also means that libCellML developers should have clang-format installed on their system so that it formatting occurs automatically.

As for whether we get the same results across all platforms, I honestly can't remember. I guess I will have to double check just before we are ready to merge this in (in case other PRs are being merged before this one).

hsorby commented:
No clang format is no running on all build slaves, only on Mac OS X machine is this currently available.

agarny committed:
Merge branch 'develop' into issue284.

agarny committed:
Added a missing API header file.

agarny committed:
Reordered some list of files.

agarny committed:
Version tests: fixed the list of sources.

agarny committed:
Configure files: make ClangFormat happy.

agarny committed:
Use PARENT_SCOPE to let ClangFormat know which files should be formatted.

agarny commented:
@hsorby and @nickerso: I believe we are now ready with this PR.

nickerso commented:
Have we confirmed consistency across platforms?

hsorby commented on src/api/libcellml/namespaces.h:
```diff
@@ -20,10 +20,10 @@ limitations under the License.
-const char * const CELLML_2_0_NS = "http://www.cellml.org/cellml/2.0#";
-const char * const MATHML_NS     = "http://www.w3.org/1998/Math/MathML";
-const char * const XLINK_NS      = "http://www.w3.org/1999/xlink";
+const char *const CELLML_2_0_NS = "http://www.cellml.org/cellml/2.0#";
```
Do we need these in the public API.  They are not used in any tests so that would indicate that we don't so this should then be a private header.

agarny replied:
Agreed. Consider it done.

hsorby commented on src/api/libcellml/printer.h:
```diff
@@ -199,4 +199,4 @@ class LIBCELLML_EXPORT Printer: public Logger
 };
 
-}
+} // namespace libcellml
```
I take it that this is driven from clang tidy?

agarny replied:
Correct. This is something that we can configure though. Personally, I like to know which namespace is being "closed", but happy to go with the majority.

hsorby replied:
If clang format is adding it for me then I have no problem with it.

hsorby commented on src/configure/versionconfig.h.in:
```diff
@@ -20,11 +20,13 @@ limitations under the License.
-static const std::string LIBCELLML_LIBRARY_VERSION_STRING="@LIBCELLML_LIBRARY_VERSION_STRING@";
+static unsigned int LIBCELLML_LIBRARY_VERSION = @LIBCELLML_LIBRARY_VERSION@;
+// clang-format on
+static const std::string LIBCELLML_LIBRARY_VERSION_STRING = "@LIBCELLML_LIBRARY_VERSION_STRING@";
```
What are the clang tidy failings here?

agarny replied:
Yes, sorry, I should have put a comment about this in my commit.

Basically, after ClangFormat has done its job,

```
static unsigned int LIBCELLML_LIBRARY_VERSION = @LIBCELLML_LIBRARY_VERSION@; 
```

will be

```
static unsigned int LIBCELLML_LIBRARY_VERSION = @LIBCELLML_LIBRARY_VERSION @;
```

i.e. there is a space before the second `@`, which in turn means that the file cannot be properly configured.

hsorby reviewed (commented):
A couple of questions from me.

hsorby commented:
I don't want to be having to get clang format onto a Windows machine if I have to get the whole of llvm for it.  Is there  a direct download for it somewhere, I have so far not been able to find one.

agarny commented:
> I don't want to be having to get clang format onto a Windows machine if I have to get the whole of llvm for it. Is there a direct download for it somewhere, I have so far not been able to find one.

No, to get ClangFormat is an all or nothing download, i.e. you need to download all of LLVM+Clang. Also, although I have LLVM+Clang 8.0 installed on my Windows machine, ClangFormat gets overlooked when the VS 2017 environment gets set. Indeed, VS 2017 comes with its own copy of ClangFormat. The problem is that it's version 6.0, so it complains about some of the things in our `.clang-format` file. So, I have kind of "deleted" VS 2017's copy of ClangFormat, so that I can use ClangFormat 8.0 on Windows.

Either way, I would argue that it's good enough to have ClangFormat tested only on one of our CI slaves. At the end of the day, it's really on the developer's computer that it is useful, not at the CI stage. That is, unless we do a `git diff` to find out what is different between the PR and the code resulting from our call to ClangFormat.

> Have we confirmed consistency across platforms?

We haven't and it's kind of difficult for me to confirm it now since all the `.cpp` and `.h` files have been properly formatted. However, I guess I could "manually" test things to confirm this. Ok, consider it done.

agarny committed:
Moved `namespaces.h` away from the API.

hsorby commented:
Agreed on only needing to have one of the build slaves have clang format.  Will have to do some work on checking if Buildbot is engaged in running the clang format check.

agarny committed:
Some minor cleaning up.

Replaced some tabs with spaces.

agarny commented:
> We haven't and it's kind of difficult for me to confirm it now since all the `.cpp` and `.h` files have been properly formatted. However, I guess I could "manually" test things to confirm this. Ok, consider it done.

My [`issue284-test` branch](https://github.com/agarny/libcellml/tree/issue284-test) includes all the ClangFormat-related code in this PR (i.e. does not include changes to our different `.cpp`/`.h` files as a result of calling ClangFormat on them). From there, I configured libCellML on Windows, Linux and macOS (which results in calling ClangFormat on all our `.cpp`/`.h` files), did a `git diff` on each platform and compared the output of those `git diff`'s. Bottom line: they are all identical. So, we do have consistency across platforms. Yeah! :) (FWIW, we may not have consistency across platforms with Clang-Tidy.)

hsorby commented:
Well since this is a consideration on clang format we can happily ignore clang tidy!

nickerso commented:
wouldn't it make more sense to go with the version that ships with the version of Visual Studio that we target? As with @hsorby, I don't want to be installing an entire tool chain that I won't use for anything else just to make sure my formatting is correct...are there any issues with clang-format 6.0 that mean we can't use it?

Is the idea that the CI would use this to reformat code before a PR is accepted or that it would be used to test for poorly formatted code in a PR and fail it in the same way as the tests, coverage, etc. ? i.e., as a "casual" developer, do I need to have clang-format on my machine or not?

agarny commented:
> wouldn't it make more sense to go with the version that ships with the version of Visual Studio that we target? As with @hsorby, I don't want to be installing an entire tool chain that I won't use for anything else just to make sure my formatting is correct...are there any issues with clang-format 6.0 that mean we can't use it?

Agreed, I would also like to be able to use VS's version of ClangFormat, but... it does indeed break things with our current `.clang-format` file. This is how I found out that VS comes with its own version of ClangFormat.

IOW, if we want to use our current `.clang-format`, we must use ClangFormat 8.0. ClangFormat 6.0 is not good enough. So, two choices:
1) Use ClangFormat 6.0 on all platforms, but this would a bit silly considering how fast ClangFormat is evolving.
2) Use ClangFormat 8.0 on all platforms, meaning that we somehow need to ensure that we get the right version on Windows (besides the fact that people probably have no choice but to install all of LLVM+Clang on Windows).

> Is the idea that the CI would use this to reformat code before a PR is accepted or that it would be used to test for poorly formatted code in a PR and fail it in the same way as the tests, coverage, etc. ? i.e., as a "casual" developer, do I need to have clang-format on my machine or not?

Ideally, people would have ClangFormat installed on their machine. However, there is no way to enforce it. So, I would expect the CI (only the macOS slave?) to do a `git diff` after ClangFormat has been performed. From there, if there are differences, the build is considered to have failed otherwise we carry on with tests, coverage and memory check.

hsorby commented:
I think the best way for us is to say if your clang-format is not version 8 (at least) then no dice you can't do the clang-format thing.

hsorby commented:
I'm working on detecting clang-format violations that will be easy for our CI.

No the casual developer won't *need* clang-format.

Yes we will have a CI test to stop ugly code getting through.

kerimoyle commented:
Coming from the outside, it would be really useful if there was a sandbox
build/test/coverage/clang format everything place so that I could see what
I needed to fix without creating a PR and then having it all fail all over
the place (despite what it does on my local machine).  If there was a way
to mimic the "real" CI processes somewhere else then people could fix their
own code first without having to worry about local installations or
versions?  Casual people could then test their contributions in private
first :)

On Wed, 8 May 2019 at 10:54, Hugh Sorby <notifications@github.com> wrote:

> I'm working on detecting clang-format violations that will be easy for our
> CI.
>
> No the casual developer won't *need* clang-format.
>
> Yes we will have a CI test to stop ugly code getting through.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cellml/libcellml/pull/310#issuecomment-490406148>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ALFXCNGLCHZMPB2BN6KV4KDPUKINBANCNFSM4HI2WMIA>
> .
>

hsorby commented:
Yeah, you can offer this with Buildbot and running what is known as a `try_scheduler`. Unfortunately it is very insecure (even for me) and it would take a lot of effort for me to implement so I am not very keen.  So unkeen that I don't think I am likely to do it.

I have no problem with people just opening a PR to trigger the CI and I feel no discomfort in chucking stuff up.  Certainly a lot less discomfort than letting people run any code they want on the build workers.

The CI testing is really not that complicated and recreating it is just a matter of launching a virtual machine (thinking about it this could be done with [Docker](https://www.docker.com/) I guess).  I went through this process recently and realised the documentation could be made a lot clearer on this.  I can't say what I was thinking when I initially wrote that.

hsorby committed:
Place clang formatting into a target.  Add a target that will test if the codebase is properly formatted (according to clang-format).

hsorby committed:
Add missing ).

hsorby committed:
Add comment to reformat_code target.

agarny committed:
Merge pull request #1 from hsorby/clang_format_284

Improvements for using clang-format.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
To make sure that our clang-format test works, i.e. it should fail because of this commit.

agarny committed:
Some minor cleaning up.

agarny commented:
Ok, commit dea2552 should have failed our tests and it doesn't, so need to do something about that.

agarny commented:
> Ok, commit [dea2552](https://github.com/cellml/libcellml/commit/dea25524cd793298b826478b7e9512dd5c191f18) should have failed our tests and it doesn't, so need to do something about that.

Ok, talking with @hsorby, everything is actually in place. It's just that Buildbot needs to do something like `make test_clang_format` to check things with regards to ClangFormat.

hsorby committed:
Print out the failed file in the error message on failure.

hsorby committed:
Merge branch 'issue284' into clang_format_284

agarny committed:
Only clang-format Git-based files.

agarny committed:
Merge pull request #2 from hsorby/clang_format_284

Print out the failed file in the error message on failure.

agarny committed:
Some minor cleaning up.

agarny committed:
Some minor cleaning up.

agarny committed:
Get our CI to report two non-properly formatted files.

agarny committed:
Report all the files that were not properly formatted.

agarny committed:
List the number of files [not] properly formatted.

agarny committed:
ClangFormat: report a relative path rather than an absolute one.

Indeed, to have an absolute path may not be very useful, especially
when it comes from Buildbot.

hsorby commented on cmake/cmake_command_check_code_formatting.cmake.in:
```diff
@@ -0,0 +1,30 @@
+else()
+  list(LENGTH _FAILED_FILES _FAILED_FILES_COUNT)
+  if(_FAILED_FILES_COUNT EQUAL 1)
+    message("1 file out of ${_FILES_COUNT} files is not properly formatted:")
```
We need the first message from the failed messages to be a FATAL_ERROR to set the return code to non-zero.

hsorby replied:
or the last.

agarny replied:
I was wondering about that. Anyway, if anything it would have to be the last one. However, this messes up the output. Will find a solution.

hsorby reviewed (changes requested):
Set return code through FATAL_ERROR message.

agarny committed:
ClangFormat: make sure that we generate a non-zero return code.

agarny commented:
I am puzzled:
- Why do the memory leaks fail now? I have just checked the output and everything is the same as when they used to pass...!?
- Why does the ClangFormat (notice no space in [ClangFormat](https://clang.llvm.org/docs/ClangFormat.html), although there doesn't seem to be a consistent spelling for it) test remain orange? FWIW, we now generate a `FATAL_ERROR` message.

hsorby commented:
Not sure on either of those two points but I'm looking into it.

* point one is now fixed that was due to an accidental copy paste.
* point two is now fixed and Buildbot is correctly reporting the build status for clang-format.

hsorby committed:
Update cmake_command_check_code_formatting.cmake.in

Setting the level of message to STATUS, to make Buildbots life easier.

agarny committed:
Make ClangFormat happy.

agarny commented:
Great, it looks to me like everything is in place now.

hsorby commented:
I think we need some documentation on the clang format capabilities that we now have before we can merge this in.

agarny commented:
> I think we need some documentation on the clang format capabilities that we now have before we can merge this in.

Fair enough. Is that something that you want to do (since, I believe, you have done most of the documentation so far?)?

nickerso commented:
Bumping the documentation to a separate issue, #315.

@hsorby, are you happy that your requested changes have all been addressed?

hsorby commented:
Sure, why not.
